src/net/ssl-context.cpp

```diff
@@ -7,43 +7,65 @@
 #include <openssl/ossl_typ.h>
 #include <openssl/ssl.h>
 #include <exception>
+#include <iostream>
+#include <measurement_kit/common.hpp>
 #include "src/net/ssl-context.hpp"
 
 namespace mk {
 namespace net {
 
-SSL_CTX *SslContext::get_client_context() {
-    static SslContext singleton;
-    return singleton.ctx;
-}
-
-SSL *SslContext::get_client_ssl() {
-    SSL *ssl = SSL_new(get_client_context());
+SSL *SslContext::get_client_ssl(std::string hostname) {
+    SSL *ssl = SSL_new(ctx);
     if (ssl == nullptr) {
+        debug("ssl: failed to call SSL_new");
         throw std::exception();
     }
+
+    SSL_set_tlsext_host_name(ssl, hostname.c_str());
     return ssl;
 }
 
-SslContext::SslContext() {
-    // TODO: check return values where applicable
+void SslContext::init(std::string ca_bundle_path) {
     SSL_library_init();
     ERR_load_crypto_strings();
     SSL_load_error_strings();
     OpenSSL_add_all_algorithms();
 
-    // FIXME: the code in here is really simple because we only aim to
-    // smoke test the OpenSSL functionality. What is missing here is at
-    // the minimum code to verify the peer certificate. Also we should
-    // not allow SSLv2 (and probably v3) connections, and we should also
-    // probably set SNI for the other end to respond correctly.
-    ctx = SSL_CTX_new(SSLv23_client_method());
+    debug("ssl: creating ssl context with bundle %s", ca_bundle_path.c_str());
+
+    ctx = SSL_CTX_new(TLSv1_client_method());
     if (ctx == nullptr) {
+        debug("ssl: failed to create SSL_CTX");
         throw std::exception();
     }
+
+    if (SSL_CTX_load_verify_locations(ctx, ca_bundle_path.c_str(), NULL) == 0) {
```

Comment:
I'd prefer if this would be `!= 1`. With C libraries I am always afraid they would agument semantic adding a novel error code in the future (like adding `-1` to distinguish from `0`), thus I prefer the code to be very defensive and basically check "if it's different from what the expected positive outcome would be".